### DISCUSSION TITLE
Trying to write `gl_FragDepth`

### DISCUSSION THREAD

**Discussion by xinaesthete** [visgl/deck.gl#7391] [Q&A]
I've been trying to adapt a volume renderer from Viv so that it writes to the depth buffer, but for some reason I don't seem to be seeing any results from calling `gl_FragDepth = depth;` in the fragment shader. As of this moment, if I was able to set `gl_FragDepth = 0.;` and see most other things being hidden, that would be good progress.

There seems to be something about the GL state that I am missing, or that I don't well enough understand how to set with `deck.gl` / `luma.gl`... I've tried using [spector.js](https://spector.babylonjs.com/) to capture the GL state as it renders the volume layer, notably:

```
Depth State
DEPTH_TEST: false
DEPTH_FUNC: LEQUAL
DEPTH_RANGE: 0, 1
DEPTH_WRITEMASK: true
disabledCommandIds: 1235
```

So, this appears to show that depth write is enabled at that point, although I'm suspicious of this `disabledCommandIds` thing which I don't understand (I briefly attempted to set up spector so that I could debug how it was deriving that information and what it might mean, but didn't quite manage). I've tried fiddling with the `parameters` GL props of my volume layer, but can't see anything that doesn't seem to be set as I think it should be. It definitely seems to be primarily a GL state problem rather than an issue with the shader code AFAICT.

With the current setup, I render the volume first with a scatterplot layer after. I want to see the scatterplot partially occluded by the volume. There may be some merit to a different approach, where other opaque geometry renders first into a depth-buffer, which the volume shader uses to discard parts of the volume that should be occluded. That would probably allow better results with transparent volume rendering in particular, but seemed like it would require more knowledge of deck/luma and boilerplate around how to pass the appropriate props between passes.... I would still like to try that approach ([shadow pass](https://github.com/visgl/deck.gl/blob/2877eb341778e04294fd95053d2ba2320ff958ce/modules/core/src/passes/shadow-pass.ts) is a starting point), but I really want to figure out why this isn't working to put my mind at rest!

You can see some work-in-progress [here](https://obvios-proto.vercel.app/) (scroll down to the third example as of this writing). I could probably share the actual code if that helps. Note that there is a toggle to enable depth-visualisation, but currently that doesn't effect the volume rendering until you select a different 'resolution' value (hint, the bottom lowest-res one will load a lot faster).

I've had some lengthy [discussion of this on the Viv repo](https://github.com/hms-dbmi/viv/discussions/649) with @ilan-gold who had also been working on similar lines earlier. I haven't quite figured out why my results aren't quite consistent with his when running from the relevant branch...

**Comment by ilan-gold** [comment]
@xinaesthete Have you been able to get this working outside of the context of Viv?  Say, on a customized `PointCloudLayer` or something via an extension to write at `fs:#main-end` or the like?

**Reply by xinaesthete** [reply]
Good question, I can't remember right now - at some point I was trying this with one of the deck.gl codepen examples as a starting point, but I don't think I picked the best layers to test with and ended up moving on before I'd properly verified.

**Reply by xinaesthete** [reply]
OK, so I just tried making a customized `ScatterplotLayer`, and it does appear that `gl_FragDepth` writing works in that context (in the same prototype as before, with the scatterplot layer just after a `VivVolumeLayer` in a `VivVolumeView`).

I tried to use a `LayerExtension` with the regular `ScatterplotLayer`, but it gets glsl compile errors writing `gl_FragDepth` because of being `#version 100`. Now realised that I can easily `extends ScatterplotLayer` to change that.

So anyway, that is certainly progress. Perhaps from here I can do a bit more `spector.js` debugging to see what changes there are in the gl depth_state or anything else that might be relevant to getting it working with the volume shader.

**Reply by ilan-gold** [reply]
So then the issue appears to be localized to our layer.  I can try to debug that specific issue a bit then.

I suspect that once this is done, we may still want want to edit the blending per `draw` call of each layer, but that is a next step.

**Reply by xinaesthete** [reply]
OK, so actually I'm trying again with your `new_3d_axes` viv branch and it is behaving more like I'd expect. I'm on a different (windows) machine now, which could be relevant, or it could be that I made a mistake when I tried to test it before (the first volume I tried to load just now didn't have any 3d axes because it didn't have `source?.meta?.physicalSizes`).

So I'll do some more testing on this machine as well as my laptop and see what I figure out. Hopefully getting there...

**Reply by xinaesthete** [reply]
ok, so it is the same with `new_3d_axes` on the other machine as well - I must have made a mistake at some point in my testing. Anyway, I'll move any viv-centric discussion back to the thread there, but maybe update here when it comes to getting to the bottom of that particular issue I seemed to be having with deck.
